xds/src/main/java/io/grpc/xds/GcpAuthenticationFilter.java

```diff
@@ -0,0 +1,226 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import com.google.auth.oauth2.ComputeEngineCredentials;
+import com.google.auth.oauth2.IdTokenCredentials;
+import com.google.common.primitives.UnsignedLongs;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.Message;
+import io.envoyproxy.envoy.extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig;
+import io.envoyproxy.envoy.extensions.filters.http.gcp_authn.v3.TokenCacheConfig;
+import io.grpc.CallCredentials;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.CompositeCallCredentials;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.auth.MoreCallCredentials;
+import io.grpc.xds.Filter.ClientInterceptorBuilder;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link Filter} that injects a {@link CallCredentials} to handle
+ * authentication for xDS credentials.
+ */
+final class GcpAuthenticationFilter implements Filter, ClientInterceptorBuilder {
+
+  static final String TYPE_URL =
+      "type.googleapis.com/envoy.extensions.filters.http.gcp_authn.v3.GcpAuthnFilterConfig";
+
+  public GcpAuthenticationFilter() {
+
+  }
+
+  @Override
+  public String[] typeUrls() {
+    return new String[] { TYPE_URL };
+  }
+
+  @Override
+  public ConfigOrError<? extends FilterConfig> parseFilterConfig(Message rawProtoMessage) {
+    GcpAuthnFilterConfig gcpAuthnProto;
+    if (!(rawProtoMessage instanceof Any)) {
+      return ConfigOrError.fromError("Invalid config type: " + rawProtoMessage.getClass());
+    }
+    Any anyMessage = (Any) rawProtoMessage;
+
+    try {
+      gcpAuthnProto = anyMessage.unpack(GcpAuthnFilterConfig.class);
+    } catch (InvalidProtocolBufferException e) {
+      return ConfigOrError.fromError("Invalid proto: " + e);
+    }
+
+    long cacheSize = 10;
+    // Validate cache_config
+    TokenCacheConfig cacheConfig = gcpAuthnProto.getCacheConfig();
+    if (cacheConfig != null) {
+      cacheSize = cacheConfig.getCacheSize().getValue();
+      if (cacheSize == 0) {
+        return ConfigOrError.fromError(
+            "cache_config.cache_size must be in the range (0, INT64_MAX)");
+      } else if (UnsignedLongs.compare(cacheSize, 0L) < 0) {
```

Comment:
Why not throw ConfigError for the negative value as well?

Reply:
gRFC says,
> cache_size: Optional. If set, must be greater than 0. Defaults to 10. Implementations that cannot support caches as large as UINT64_MAX may cap this value at their maximum supported size.

The caller should be using this with unsigned values. Negative would mean very large in uint, so we are simply making the size to our maximum.

Reply:
I think the important thing to realize is this is impossible. You can't have negative values for an unsigned long, so this condition always fails and we should remove it.

Reply:
So I think we should not use `UnsignedLongs.compare(cacheSize, 0L) < 0`, instead we should use `cacheSize < 0`. If so, then we should make `cacheSize = Interger.MAX_VALUE-1`?

Reply:
This is not at all clear _why_ the behavior is that way. Using `UnsignedLongs` makes it much more clear, and simpler:
```java
if (UnsignedLongs.compare(cacheSize, Integer.MAX_VALUE) >= 0) {
  cacheSize = integer.MAX_VALUE - 1;
}
// Or even clearer:
cacheSize = UnsignedLongs.min(cacheSize, Integer.MAX_VALUE - 1);
```

If we are going to have the MAX_VALUE - 1 in the config processing, it should have a comment why the -1 is there.

And we should also store an int in the config object instead of a long, if it is guaranteed to be an int.

Reply:
Seems fine. Generally I wouldn't say "Maximum size should not go beyond Integer.MAX_VALUE-1" because that's clear in the code. "LruCache's size is an int and briefly exceeds its maximum size before evicting entries" would be fine too. Anything hinting at "LruCache" and "overflow" seems fine.